pkg/crypto/signer.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package crypto
+
+import (
+	"crypto/ecdsa"
+
+	"github.com/btcsuite/btcd/btcec"
+)
+
+type Signer interface {
+	Sign(data []byte) ([]byte, error)
+	PublicKey() (*ecdsa.PublicKey, error)
+}
+
+type Recoverer interface {
+	Recover(signature, data []byte) (*ecdsa.PublicKey, error)
+}
+
+// Recover verifies signature with the data base provided.
+// Is exported so it can be used in other `Recoverer` implementations as well.
+func Recover(signature, data []byte) (*ecdsa.PublicKey, error) {
+	p, _, err := btcec.RecoverCompact(btcec.S256(), signature, data)
+	return (*ecdsa.PublicKey)(p), err
+}
+
+type SignRecoverer interface {
```

Comment:
Still don't like this merged interface. To me this makes the impression that the `Recover` operation depends on the signer you yourself use when really it depends on the signer the other peer uses (which we cannot know and which really should produce the signature in the same format as the `defaultSigner` anyway).

I think `Recoverer` and `Signer` should be passed to the service separately and there should only be a single implementation of `Recoverer` (at least as long as no information on the signature format is passed as part of the handshake).

Reply:
I think in this case this is only because of the code, not to use 2 components. So you have basically 3 options, to use `Signer`, `Recoverer`, and `SignRecoverer` if you need both, but there is nothing stopping you to implement them however you want. This is more like a helper abstraction for the code. There is no dependency between `Signer` and `Recoverer` imo, it's just extracted in the interface. LTMWYT

Reply:
Well, if we are not going to have multiple recover functions, I can completely delete `Recoverer` interface and use `crypto.Recover()` function directly in the code as well. I don't mind, no need to make abstraction if we don't need them and the function is small enough to directly use in unit tests as well. @janos what do you say as well? This way we will only have `Signer` interface

Reply:
We can do that as we only use one elliptic curve. Does it make sense to make it explicit than like `RevoverSecp256k1`?

Reply:
If we put it into the name it should probably be more explicit about the signature format. There are many different formats for `secp256k1` based signatures. This recover function would be specifically for compact signatures.

One case where `Recoverer` might make sense is if we allow different schemes / formats for the same curve. E.g. you could let your peer know that you sign data with the "Ethereum Signed Message"-Prefix and then it would use a different `Recoverer` for you. This would allow things like `clef` for external signing also for overlay but would require transmitting this information in the handshake.

Reply:
I would go for a single `crypto.Recover()` then, as this way it is not self-explanatory, so one who is using it will have to take a look at the code, but I don't see a way to overcome this. At least, it is obvious there is a single `Recover` function. But I think we are kind of over-engineering it for now, if we stumble upon different `Recover` functions, we can change the name, add the interface, do whatever we like :) 
I thought that we said we won't allow different formats nor curves. Actually, in the beginning of conversions, I used format and later we switched to curve :) WDYGT?